File: tests/envs/llvm/service/ActionSpaceTest.cc
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.
#include <gtest/gtest.h>

#include <magic_enum.hpp>

#include "compiler_gym/envs/llvm/service/ActionSpace.h"
#include "tests/TestMacros.h"

using namespace ::testing;

namespace compiler_gym::llvm_service {
namespace {

TEST(ActionSpacesTest, getLlvmActionSpace) {
  const auto spaces = getLlvmActionSpaceList();
  EXPECT_EQ(spaces.size(), 1);
  EXPECT_EQ(spaces[0].name(), "PassesAll");
  EXPECT_EQ(spaces[0].action_size(), magic_enum::enum_count<LlvmAction>());
}

}  // anonymous namespace
}  // namespace compiler_gym::llvm_service

File: tests/llvm/action_space_test.py
# Copyright (c) Facebook, Inc. and its affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
"""Tests for the LLVM environment action space."""
from compiler_gym.envs.llvm.llvm_env import LlvmEnv
from tests.test_main import main

pytest_plugins = ["tests.pytest_plugins.llvm"]


def test_to_and_from_string_no_actions(env: LlvmEnv):
    env.reset(benchmark="cbench-v1/crc32")
    assert env.action_space.to_string(env.actions) == "opt  input.bc -o output.bc"
    assert env.action_space.from_string(env.action_space.to_string(env.actions)) == []


def test_to_and_from_string(env: LlvmEnv):
    env.reset(benchmark="cbench-v1/crc32")
    env.step(env.action_space.flags.index("-mem2reg"))
    env.step(env.action_space.flags.index("-reg2mem"))
    assert (
        env.action_space.to_string(env.actions)
        == "opt -mem2reg -reg2mem input.bc -o output.bc"
    )
    assert env.action_space.from_string(env.action_space.to_string(env.actions)) == [
        env.action_space.flags.index("-mem2reg"),
        env.action_space.flags.index("-reg2mem"),
    ]


if __name__ == "__main__":
    main()
